envelope.go

```diff
@@ -40,248 +38,168 @@ var errSpecialEnvelope = errorf(
 // message length. gRPC and Connect interpret the bitwise flags differently, so
 // envelope leaves their interpretation up to the caller.
 type envelope struct {
-	Data  *bytes.Buffer
-	Flags uint8
+	Buffer *bytes.Buffer
 }
 
-func (e *envelope) IsSet(flag uint8) bool {
-	return e.Flags&flag == flag
+// makeEnvelope creates an envelope with a 5-byte prefix.
+func makeEnvelope(buffer *bytes.Buffer) envelope {
+	var head [5]byte
+	buffer.Write(head[:])
+	return envelope{Buffer: buffer}
```

Comment:
With these changes, is `envelope` doing enough work to keep around? It doesn't seem to provide much value as a wrapper on top of `bytes.Buffer`, and I don't love that the `unwrap` method isn't safe to call twice.

Would the rest of the repository be more clear if we instead had something like this (and managed the envelope and message buffer separately)?

```go
type envelope [5]byte

func (e *envelope) Flags() uint8 { ... }
func (e *envelope) SetFlags(uint8) { ... }
func (e *envelope) HasFlags(uint8) { ... } // check if flags are set
func (e *envelope) Size() int { ... } // read size from last 4 bytes
func (e *envelope) SetSize(int) { ... } // write size to last 4 bytes
```